### PR TITLE
Revert "Fix gradient checkpointing + fp16 autocast for most models"

### PR DESCRIPTION
**Pull request by younesbelkada**
Reverts huggingface/transformers#24247

This PR reverts #24247 

The investigation initially started with the failing test in https://github.com/huggingface/peft/actions/runs/5340918925/jobs/9686171926 - a training setup that was taking 7GB now takes 15GB and OOM. I looked back at each commit and can confirm this commit caused it.  

Instead of patching the initial issue on our side, I propose for now to revert the PR and just wait for the fix in PT side as doubling down the memory requirements is a lot for PEFT users. 

Can confirm the training doesn't OOM before the commit 285a48011da3145ae77c5b22bcfbe77d367e5173 hence the PR that reverts the commit

cc @sgugger @pacman100 @amyeroberts  

Putting it as draft as I need to deep dive a bit before making sure this is the right thing to do

### CODE REVIEW

**Comment by amyeroberts** [comment]
I'd be pro reverting until we manage to resolve or find another solution. 

cc @ydshieh too as the testing master :) Looking at our daily CI, it seems this hasn't affected our "normal" models - is this right? Are there any tests we should be running to verify this?

**Comment by ydshieh** [comment]
@amyeroberts 

That PR #24247 is merged yesterday. The daily CI is triggered this morning and not finished yet. So we don't know what that PR brings.

**Comment by ydshieh** [comment]
There is also a push CI (only non-slow tests and not a complete CI)

Screenshot 2023-06-22 114431

**Comment by younesbelkada** [comment]
Ok I just did some benchmarks by observing the peak memory usage of different training setups and it seems to affect most of the models regardless of the modality:

| Model  | Quantization method  | Use Rentrant (i.e. #24247 included) | Peak memory usage |
| -------- | ------- | ------- | ------- |
| `openai/whisper-large`   |  8bit | Yes  | OOM |
| `openai/whisper-large`   |  8bit | No | 7.5GB |
| `openai/whisper-large`   |  4bit | No | 5.1GB |
| `openai/whisper-large`  |  4bit | Yes | 14.5GB |
| `facebook/opt-6.7b`  |  8bit | Yes | 14.1GB |
| `facebook/opt-6.7b` |  8bit | no | 9.8GB |

Note that before #24420 the last PEFT layer had always None grad, therefore got never updated. But the surprising thing is that the last layer shouldn't cause 2x memory increase

I will investigate further and keep updates here

**Comment by amyeroberts** [comment]
@younesbelkada Thanks for investigating and sharing! Could you also add a model with no quantization for reference in the table?

**Comment by younesbelkada** [comment]
Sure yes! Will update the table soon

**Comment by younesbelkada** [comment]
From the updated observations above

1- it seems to affect the quantized models only
2- Larger models gets more affected

**Comment by younesbelkada** [comment]
Perfect, let's revert this PR then 
I can also confirm I don't have any None-grad for lora layers using llama (as posted in original issue), I believe the recent accelerate integration silently fixed the bug and the user was using a former version of transfomers

cc @amyeroberts @sgugger this is ready for review

**Review by amyeroberts — APPROVED** [review]
Revert LGTM. Thanks for acting so quickly on this and the detailed investigation!

**Comment by younesbelkada** [comment]
Thanks very much for the support and quick feedback! @amyeroberts
